### ISSUE TITLE
Verify BFLOAT16 Addition in FPU

### ISSUE THREAD

**Issue by KalaivaniMCW** [tenstorrent/tt-metal#34664]
There are some mismatches observed when exhaustively testing BFLOAT16 addition op resulting in ULP 1

Pytest to repro: 
```
def test_binary_bf16(device):
    torch.manual_seed(0)
    torch_dtype = torch.bfloat16
    ttnn_dtype = ttnn.bfloat16

    x_torch = torch.tensor([[0.00341796875, 0.003448486328, 0.00373840332, 0.01129150391, 0.01953125]], dtype=torch_dtype)
    y_torch = torch.tensor([[1, 1, 1, 1, 1]], dtype=torch_dtype)
    z_torch = torch.add(x_torch, y_torch)
    x_tt = ttnn.from_torch(x_torch, dtype=ttnn_dtype, layout=ttnn.TILE_LAYOUT, device=device)
    y_tt = ttnn.from_torch(y_torch, dtype=ttnn_dtype, layout=ttnn.TILE_LAYOUT, device=device)
    z_tt = ttnn.add(x_tt, y_tt)
    tt_out = ttnn.to_torch(z_tt)

    torch.set_printoptions(linewidth=200, threshold = 10000 , precision=15, sci_mode = False, edgeitems=17)
    print("z_tt", z_tt)
    print("tt_out", tt_out)
    print("z_torch", z_torch)

    assert_with_ulp(z_torch, tt_out, 1)
```
Torch vs TT outputs 
```
TT output as ttnn tensor 
z_tt ttnn.Tensor([[ 1.0078,  1.0078,  1.0078,  1.0156,  1.0234]], shape=Shape([1, 5]), dtype=DataType::BFLOAT16, layout=Layout::TILE)

TT output after to_torch 
tt_out tensor([[1.007812500000000, 1.007812500000000, 1.007812500000000, 1.015625000000000, 1.023437500000000]], dtype=torch.bfloat16)

Torch output
z_torch tensor([[1.000000000000000, 1.000000000000000, 1.000000000000000, 1.007812500000000, 1.015625000000000]], dtype=torch.bfloat16)
```

The LLK api executed here is add_tiles which leads to this implementation 
https://github.com/tenstorrent/tt-llk/blob/0668e99c6af0017db4ebd952dc862edeb20077e2/tt_llk_wormhole_b0/llk_lib/llk_math_eltwise_binary.h#L42

Are these results expected for addition in FPU ? 
Is there any rounding adjustment that could be made to get a more precise result ?

**Comment by KalaivaniMCW** [comment]
<section id="markdown-section-fc9e7bc9-ad7a-4717-9585-eac62e15939f-1" class="markdown-section chat-fade-in" data-markdown-raw="

## Test Results Table for `test_binary_bf16` (bfloat16 addition)" data-section-index="1" style="border-radius: 4px; line-height: 19.5px; margin: 6px 0px; position: relative; scroll-margin-bottom: 40px; scroll-margin-top: 40px; color: rgba(228, 228, 228, 0.92); font-family: -apple-system, &quot;system-ui&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(20, 20, 20); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h2 style="animation: 0.25s ease 0s 1 normal none running fade-in; font-weight: 600 !important; font-size: 1.3em; line-height: 1.25; margin-bottom: 10px; margin-top: 20px;"><span style="animation: 0.25s ease 0s 1 normal none running fade-in;">Test Results Table for<span> </span></span><span class="markdown-inline-code leading-[1.4]" style="animation: 0.25s ease 0s 1 normal none running fade-in; line-height: 1.4; background-color: rgba(10, 10, 10, 0.4); border-radius: 4px; color: rgb(129, 161, 193); font-family: Menlo, Monaco, &quot;Courier New&quot;, monospace, Menlo, Monaco, &quot;Courier New&quot;, monospace; font-size: 0.9em; padding: 1.5px 3px; transition: 0.1s; border-color: rgb(129, 161, 193); word-break: break-all; cursor: pointer;"><span style="animation: 0.25s ease 0s 1 normal none running fade-in;">test_binary_bf16</span></span><span style="animation: 0.25s ease 0s 1 normal none running fade-in;"><span> </span>(bfloat16 addition)</span></h2></section><div class="monaco-scrollable-element  mac" role="presentation" style="scrollbar-width: none; overflow: hidden; color: rgba(228, 228, 228, 0.92); font-family: -apple-system, &quot;system-ui&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(20, 20, 20); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; position: relative; margin: 1em 0px; width: 816px;"><div style="overflow: hidden;">
Index | Input X | Input Y | Expected (torch) | Actual (ttnn) | Match? | Notes
-- | -- | -- | -- | -- | -- | --
0 | 0.00341796875 | 1.0 | 1.0000000000 | 1.0078125000 | ❌ | Off by ~0.0078
1 | 0.003448486328 | 1.0 | 1.0000000000 | 1.0078125000 | ❌ | Off by ~0.0078
2 | 0.00373840332 | 1.0 | 1.0000000000 | 1.0078125000 | ❌ | Off by ~0.0078
3 | 0.01129150391 | 1.0 | 1.0078125000 | 1.0156250000 | ❌ | Off by ~0.0078
4 | 0.01953125 | 1.0 | 1.0156250000 | 1.0234375000 | ❌ | Off by ~0.0078
5 | 0.1 | 1.0 | 1.1015625000 | 1.1015625000 | ✅ | Exact match
6 | 0.001 | 1.0 | 1.0000000000 | 1.0000000000 | ✅ | Exact match
7 | 0.0001 | 1.0 | 1.0000000000 | 1.0000000000 | ✅ | Exact match

</div><div role="presentation" aria-hidden="true" class="invisible scrollbar horizontal fade" style="visibility: hidden; opacity: 0; pointer-events: none; transition: opacity 0.8s linear; position: absolute; width: 401px; height: 10px; left: 0px; bottom: 0px;">

Observations:

1. **Pattern**: TTNN output is consistently ~0.0078125 (1/128 in bfloat16) higher than expected for indices 0-4
2. **Small values work**: Very small additions (0.001, 0.0001) round correctly to 1.0
3. **Larger value works**: 0.1 + 1.0 = 1.1015625 is exact
4. **Problem range**: Values in ~0.003-0.02 range show consistent +0.0078125 bias
This suggests a potential rounding or precision issue in TTNN's addition implementation for small values in bfloat16 when adding to 1.0.

**Comment by KalaivaniMCW** [comment]
**Explanation: for + 1 cases** 

**Rounding Mode Analysis**
**What's happening:**
In `bfloat16`, around the value `1.0`:
Representable values are spaced by 1/128 = 0.0078125 (this is the ULP at magnitude 1.0)
So the nearest representable values are: `1.0, 1.0078125, 1.015625, 1.0234375, ...`
`For index 0: 0.00341796875 + 1.0 = 1.00341796875`
This exact result is between `1.0` and `1.0078125`:
```
PyTorch (round-to-nearest-even): 1.00341796875 is closer to 1.0 → rounds down to 1.0 ✓
TTNN (round-up mode): 1.00341796875 rounds up to 1.0078125
```

`For index 3: 0.01129150391 + 1.0 = 1.01129150391`
This is between `1.0078125` and `1.015625`:
```
PyTorch: 1.01129150391 is closer to 1.0078125 → rounds to 1.0078125 ✓
TTNN: rounds up to 1.015625
```

**Evidence:**
✅ Consistent +1 ULP bias: Every mismatch shows TTNN result is exactly one ULP higher than torch
✅ Pattern matches: This is exactly what you'd see with "round towards +∞" vs "round to nearest"
✅ Larger values work: When the sum is already representable (like `1.1`), no rounding needed → match!

**IEEE 754 Rounding Modes:**
Round to nearest, ties to even (default, used by PyTorch) ← standard
Round towards +∞ (ceiling) ← likely what TTNN is using
Round towards -∞ (floor)
Round towards zero (truncate)

This is a hardware/backend rounding mode configuration issue in TTNN! 🎯

**Comment by KalaivaniMCW** [comment]
**Bfloat16 Addition with  (-1)**

BFLOAT16 representable values near -1.0: `..., -1.0078125, -1.0000000  (zero point),  -0.99609375, -0.9921875, -0.9843750, ...`
For values in [-2.0, -1.0]: exponent = 0, spacing = 1/128 = 0.0078125
For values in [-1.0, -0.5): exponent = -1, spacing = 1/256 = 0.00390625

```
Input:
x_torch = [0.1, 0.001, 0.0001, 0.001968383789, 0.002014160156, 0.00212097168, 0.00390625, -0.9960938]
y_torch = [-1, -1, -1, -1, -1, -1, -1, 0]
```

<img width="885" height="337" alt="Image" src="https://github.com/user-attachments/assets/86b8f363-a20e-496a-9e83-38c3f8856cd6" />

**Detailed Analysis of Mismatches (Indices 3, 4, 5):**

These three cases all involve adding a small positive value to -1, where the true mathematical result falls very close to the midpoint between two representable bfloat16 values:
```
-0.996093750000000 (bfloat16: 0xBF7F, binary: 1 01111110 1111111)
-1.000000000000000 (bfloat16: 0xBF80, binary: 1 01111111 0000000)
```
The midpoint between these values is -0.998046875.
**Index 3: 0.001968383789 + (-1)**
```
Input 0.001968383789 in bfloat16 → 0.001953125 (exact representation: 1/512)
True result: 0.001953125 + (-1) = -0.998046875 (exactly at midpoint)
PyTorch rounds to: -0.996093750000000 (rounds toward zero / away from -∞)
TTNN rounds to: -1.000000000000000 (rounds away from zero / toward -∞)
```

**Index 4: 0.002014160156 + (-1)**
```
Input 0.002014160156 in bfloat16 → 0.00201416015625 (exact representation)
True result: 0.00201416015625 + (-1) = -0.99798583984375
The bfloat16 neighbors are -0.996093750000000 and -1.000000000000000
-0.99798583984375 is slightly closer to the midpoint but still below it
PyTorch rounds to: -0.996093750000000 (rounds toward zero)
TTNN rounds to: -1.000000000000000 (rounds away from zero)
```

**Index 5: 0.00212097168 + (-1)**
```
Input 0.00212097168 in bfloat16 → 0.0021209716796875 (exact representation)
True result: 0.0021209716796875 + (-1) = -0.9978790283203125
This is closer to -1.0 than to -0.996093750000000, but still rounds differently
PyTorch rounds to: -0.996093750000000 (rounds toward zero)
TTNN rounds to: -1.000000000000000 (rounds away from zero)
```

**Conclusion:**
The pattern is consistent across all three mismatches: when adding a small positive value to -1 (resulting in values near -1), PyTorch rounds toward zero while TTNN rounds away from zero. This confirms the rounding mode difference:
- PyTorch bfloat16: Uses "Round Toward Zero" for negative results
- TTNN bfloat16: Uses "Round Away from Zero" for negative results

This is particularly evident in Index 3, where the result is exactly at the tie-breaking midpoint, and the two implementations choose opposite neighbors.

**Comment by KalaivaniMCW** [comment]
Related : https://github.com/tenstorrent/tt-metal/issues/30147

**Comment by KalaivaniMCW** [comment]
Difference in rounding behaviour depending on the sign of inputs: 
```
x_torch = torch.tensor([[ 0.00341796875, -0.00341796875, 0.00341796875, -0.00341796875, 0.01129150391, -0.01129150391, 0.01129150391, -0.01129150391,  ]], dtype=torch_dtype)
y_torch = torch.tensor([[1, 1, -1, -1, 1, 1, -1, -1,]], dtype=torch_dtype)

z_tt ttnn.Tensor([[ 1.0078,  0.9961, -0.9961, -1.0078,  1.0156,  0.9883, -0.9883, -1.0156]], shape=Shape([1, 8]), dtype=DataType::BFLOAT16, layout=Layout::TILE)
tt_out tensor([[ 1.007812500000000,  0.996093750000000, -0.996093750000000, -1.007812500000000,  1.015625000000000,  0.988281250000000, -0.988281250000000, -1.015625000000000]], dtype=torch.bfloat16)
z_torch tensor([[ 1.000000000000000,  0.996093750000000, -0.996093750000000, -1.000000000000000,  1.007812500000000,  0.988281250000000, -0.988281250000000, -1.007812500000000]], dtype=torch.bfloat16)
```

Bfloat16 spacing:
- For values in [±1.0, ±2.0): spacing = 1/128 = 0.0078125
- For values in [±0.5, ±1.0): spacing = 1/256 = 0.00390625

<img width="1025" height="318" alt="Image" src="https://github.com/user-attachments/assets/aaca3d4a-caaa-408a-a640-3506f53bba94" />

TTNN bfloat16 addition behavior:
✅ When operands have OPPOSITE signs (effective subtraction):
TTNN matches PyTorch RNE perfectly
Cases 1, 2, 5, 6: All match

❌ When operands have SAME sign (true addition):
TTNN rounds AWAY FROM ZERO
Cases 0, 3, 4, 7: All diverge
Positive + Positive → rounds UP (to larger magnitude)
Negative + Negative → rounds DOWN toward -∞ (to larger magnitude)

Summary:
Subtraction path (opposite signs): Uses correct RNE rounding
Addition path (same signs): Uses "Round Away from Zero" mode
